src/tracing/agent.h

```diff
@@ -31,24 +32,51 @@ class TracingController : public v8::platform::tracing::TracingController {
   }
 };
 
+class Agent;
 
-class Agent {
+// Allows agent to be destroyed early. Methods are noop after the agent was
+// destroyed
+class AgentHandle {
  public:
-  // Resetting the pointer disconnects client
-  using ClientHandle = std::unique_ptr<std::pair<Agent*, int>,
-                                       void (*)(std::pair<Agent*, int>*)>;
+  explicit AgentHandle(Agent*);
+  void Reset();
```

Comment:
Maybe call Reset() from the destructor so you don't have to call it explicitly in `Agent::~Agent()`?

Reply:
Here, AgentHandle is supposed to be a weak pointer - it allows accessing the tracing agent but does not prevent its destruction. I had a naive implementation with shared_ptr/weak_ptr but the issue there is that it may cause Agent to be deleted on a wrong thread.